web_external/js/views/map/MapAdapters.js

```diff
@@ -0,0 +1,212 @@
+minerva.core = minerva.core || {};
+
+(function () {
+    function AdapterRegistry() {
+        this.registry = {};
+        this.register = function (type, definition) {
+            this.registry[type] = definition;
+        };
+        this._createRepresentation = function (container, dataset, layerType, mapping) {
+            if (layerType === null || !_.has(this.registry, layerType)) {
+                console.error('This dataset cannot be adapted to a map layer of type [' + layerType + '].');
+                this.trigger('m:map_adapter_error', dataset, layerType);
+                return;
+            } else {
+                var adapter = this.registry[layerType];
+                var layerRepr = _.extend(new adapter(), Backbone.Events);
+                dataset.once('m:dataset_geo_dataLoaded', function () {
+                    layerRepr.once('m:map_layer_renderable', function (layer) {
+                        this.trigger('m:map_adapter_layerCreated', layer);
+                    }, this).once('m:map_layer_error', function (layer) {
+                        this.trigger('m:map_adapter_layerError', layer);
+                    }, this).initLayer(container, dataset, dataset.get('geoData'), mapping);
+                }, this).loadGeoData();
+                //
+                // Instead of dataset.loadGeoData, ideally something like
+                // girder.RestRequest({
+                //     type: "POST"
+                //     url: "adapter/" + dataset._id + "/" + layerType.toString()(),
+                //     params: userInput,
+                // }).success(function (data){
+                //     createLayer
+                // });
+            }
+        };
+    }
+    minerva.core.AdapterRegistry = _.extend(new AdapterRegistry(), Backbone.Events);
+})();
+
+minerva.rendering = minerva.rendering || {};
+minerva.rendering.geo = minerva.rendering || {};
+minerva.rendering.geo.defineMapLayer = function (layerType, layerDefinition, parentDefinition) {
+    if (parentDefinition) {
+        layerDefinition.prototype = new parentDefinition();
+    }
+    minerva.core.AdapterRegistry.register(layerType, layerDefinition);
+    return layerDefinition;
+}
+
+minerva.rendering.geo.MapRepresentation = minerva.rendering.geo.defineMapLayer('map', function () {
+    this.deleteLayer = function (container) {
+        container.deleteLayer(this.geoJsLayer);
+    },
+
+    this.setOpacity = function (opacity) {
+        this.geoJsLayer.opacity(opacity);
+    },
+
+    this.render = function (container) {
+        container.renderMap();
+    }
+});
+
+minerva.rendering.geo.GeometryRepresentation = minerva.rendering.geo.defineMapLayer('geojson', function () {
+    this.readerType = 'jsonReader',
+
+    this.initLayer = function (container, dataset, data, visProperties) {
```

Comment:
We may need exit method as well (which will be called when container gets deleted for example).